tempodb/encoding/vparquet3/block_autocomplete.go

```diff
@@ -0,0 +1,869 @@
+package vparquet3
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/grafana/tempo/pkg/parquetquery"
+	v1 "github.com/grafana/tempo/pkg/tempopb/trace/v1"
+	"github.com/grafana/tempo/pkg/traceql"
+	"github.com/grafana/tempo/tempodb/backend"
+	"github.com/grafana/tempo/tempodb/encoding/common"
+	"github.com/parquet-go/parquet-go"
+	"github.com/pkg/errors"
+)
+
+func (b *backendBlock) FetchTagValues(ctx context.Context, req traceql.AutocompleteRequest, cb traceql.AutocompleteCallback, opts common.SearchOptions) error {
+	err := checkConditions(req.Conditions)
+	if err != nil {
+		return errors.Wrap(err, "conditions invalid")
+	}
+
+	if len(req.Conditions) <= 1 { // Last check. No conditions, use old path. It's much faster.
+		return b.SearchTagValuesV2(ctx, req.TagName, common.TagCallbackV2(cb), opts)
+	}
+
+	pf, _, err := b.openForSearch(ctx, opts)
+	if err != nil {
+		return err
+	}
+
+	iter, err := autocompleteIter(ctx, req, pf, opts, b.meta.DedicatedColumns)
+	if err != nil {
+		return errors.Wrap(err, "creating fetch iter")
+	}
+
+	// TODO: The iter shouldn't be exhausted here, it should be returned to the caller
```

Comment:
I'm leaning towards the callback approach, since it's how the other tag values functions (`SearchTagValues, `SearchTagValuesV2`) do it. It keeps the code simpler I think.